toolchain/lc-compile/src/literal.c

```diff
@@ -87,6 +161,166 @@ int MakeIntegerLiteral(const char *p_token, long *r_literal)
     return 1;
 }
 
+// Match 0 | [1-9][0-9]*
+static int
+__MatchInt(const char **x_token_ptr, int* r_found)
+{
+	// If there are no digits, there is no int.
+	if (!isdigit(*(*x_token_ptr)))
+	{
+		*r_found = 0;
+		return 1;
+	}
+	
+	// If there is zero before a digit, it is malformed.
+	if ((*x_token_ptr)[0] == '0' &&
+		isdigit((*x_token_ptr)[1]))
+	{
+		return 1;
+	}
+	
+	*x_token_ptr += 1;
+	
+	// Match [0-9]*
+	while(isdigit(*x_token_ptr[0]))
+	{
+		*x_token_ptr += 1;
+	}
+	
+	*r_found = 1;
+	
+	return 1;
+}
+
+// Match [0-9]+
+static int
+__MatchZeroesInt(const char **x_token_ptr, int* r_found)
+{
+	// If there are no digits, there is no int.
+	if (!isdigit(*x_token_ptr[0]))
+	{
+		*r_found = 0;
+		return 1;
+	}
+	
+	// Match [0-9]*
+	while(isdigit(*x_token_ptr[0]))
+	{
+		*x_token_ptr += 1;
+	}
+	
+	*r_found = 1;
+	
+	return 1;
+}
+
+int IsDoubleLiteral(const char *p_token)
+{
+	// We check the pattern against:
+	//   I:Int P:Period F:ZeroesInt E:(Exp PlusMinus Int)
+	// Where the valid combinations are:
+	//   I+P
+	//   I+E
+	//   I+P+E
+	//   I+P+F
+	//   I+P+F+E
+	//   P+F
+	//   P+F+E
+	
+	int t_found_int = 0;
+	int t_found_point = 0;
+	int t_found_frac = 0;
+	int t_found_exp = 0;
+	int t_found_bits = 0;
+	const char *t_token_ptr = p_token;
+
+	// Match I - if the Int is malformed, its not a double.
+	if (!__MatchInt(&t_token_ptr, &t_found_int))
+	{
+		return 0;
+	}
+	
+	// Match P
+	if (*t_token_ptr == '.')
+	{
+		t_found_point = 1;
+		t_token_ptr++;
+	}
+	
+	// Match F - if the ZeroesInt is malformed, its not a double
+	if (!__MatchZeroesInt(&t_token_ptr, &t_found_frac))
+	{
+		return 0;
+	}
+	
+	// Match E
+	if (*t_token_ptr == 'e' || *t_token_ptr == 'E')
+	{
+		t_token_ptr++;
+		if (*t_token_ptr == '+' || *t_token_ptr == '-')
+		{
+			t_token_ptr++;
+		}
+		if (!__MatchInt(&t_token_ptr, &t_found_exp))
+		{
+			return 0;
+		}
+	}
+	
+	// If we aren't at end of the token, its not a double.
+	if (*t_token_ptr != '\0')
+	{
+		return 0;
+	}
+	
+	// Now check valid combinations by computing it as a bitset
+	t_found_bits =
+		(t_found_int << 3) |
+		(t_found_point << 2) |
+		(t_found_frac << 1) |
+		(t_found_exp << 0);
+	
+	#define I 8
+	#define P 4
+	#define F 2
+	#define E 1
+	switch(t_found_bits)
+	{
+		case I+P:
+		case I+E:
+		case I+P+E:
+		case I+P+F:
+		case I+P+F+E:
+		case P+F:
+		case P+F+E:
+			break;
+		
+		default:
+			return 0;
+	}
+	
+	/*if (!((t_found_int && t_found_point) ||
+		  (t_found_int && t_found_exp) ||
+		  (t_found_int && t_found_point && t_found_exp) ||
+		  (t_found_int && t_found_point && t_found_frac) ||
+		  (t_found_int && t_found_point && t_found_frac && t_found_exp) ||
+		  (t_found_point && t_found_frac) ||
+		  (t_found_point && t_found_frac && t_found_exp)))
+	{
+		return 0;
+	}*/
```

Comment:
Can this code block be removed, or does it need to be retained for explanatory purposes?